### PR TITLE
Feat: Add landmark attention

### LINKED ISSUE #109: [Feature] Add Landmark attention
Per request from Discord to add this feature.

They seem to monkeypatch LlamaModel and add special `mem` token.

Ref: https://github.com/epfml/landmark-attention/blob/main/llama/train.py

### PR DESCRIPTION
**Pull request by NanoCode012**
Closes #109 

NOTE: Only Llama!

- [x] Test it runs
- [x] Run tests from landmark

### CODE REVIEW

**Comment by NanoCode012** [comment]
Testing via https://github.com/epfml/landmark-attention/blob/main/llama/run_test.py

<details>

<summary>n=1k</summary>

```
Running test 50/50 for n = 1000...
Number of tokens in this prompt:  338
Expected number in the prompt: 16835, base output: 16835
Success!
Number of tokens in this prompt:  338
Expected number in the prompt: 16835, mem output: 16835
Success!
Accuracy base for n = 1000: 100.0%
Accuracy mem for n = 1000: 100.0%
```

</details>

<details>

<summary>n=5k</summary>

```
Running test 50/50 for n = 5000...
Number of tokens in this prompt:  1406
Expected number in the prompt: 18704, base output: 18704
Success!
Number of tokens in this prompt:  1406
Expected number in the prompt: 18704, mem output: 18704
Success!
Accuracy base for n = 5000: 100.0%
Accuracy mem for n = 5000: 96.0%
```

</details>

<details>

<summary>n=8k</summary>

```
Running test 50/50 for n = 8000...
Number of tokens in this prompt:  2206
Expected number in the prompt: 38907, base output: 38907
Success!
Number of tokens in this prompt:  2206
Expected number in the prompt: 38907, mem output: 38907
Success!
Accuracy base for n = 8000: 98.0%
Accuracy mem for n = 8000: 96.0%
```

</details>

<details>

<summary>n=10k</summary>

```
Running test 50/50 for n = 10000...
Number of tokens in this prompt:  2738
Expected number in the prompt: 1333, base output: .
.
.
.
.

Fail.
Number of tokens in this prompt:  2738
Expected number in the prompt: 1333, mem output: 1333
Success!
Accuracy base for n = 10000: 0.0%
Accuracy mem for n = 10000: 100.0%
```

</details>

<details>

<summary>n=12k</summary>

```
Running test 50/50 for n = 12000...
Number of tokens in this prompt:  3272
Expected number in the prompt: 10145, base output:  and and and and and
Fail.
Number of tokens in this prompt:  3272
Expected number in the prompt: 10145, mem output: 10145
Success!
Accuracy base for n = 12000: 0.0%
Accuracy mem for n = 12000: 98.0%
```

</details>

<details>

<summary>n=14k</summary>

```
Running test 10/10 for n = 14000...
Number of tokens in this prompt:  3805
Expected number in the prompt: 15978, mem output: 15978
Success!
Accuracy mem for n = 14000: 90.0%
```

</details>

<details>

<summary>n=18k</summary>

```
Running test 10/10 for n = 18000...
Number of tokens in this prompt:  4872
Expected number in the prompt: 48396, mem output: 48396
Success!
Accuracy mem for n = 18000: 100.0%
```

</details>

<details>

<summary>n=20k</summary>

```
Running test 10/10 for n = 20000...
Number of tokens in this prompt:  5406
Expected number in the prompt: 41403, mem output: 41403
Success!
Accuracy mem for n = 20000: 100.0%
```

</details>

<details>

<summary>n=25k</summary>

```
Running test 10/10 for n = 25000...
Number of tokens in this prompt:  6738
Expected number in the prompt: 30430, mem output: 30430
Success!
Accuracy mem for n = 25000: 100.0%
```

</details>

<details>

<summary>n=38k</summary>

```
Running test 10/10 for n = 38000...
Number of tokens in this prompt:  10207
Expected number in the prompt: 27396, mem output: 27396
Success!
Accuracy mem for n = 38000: 100.0%
```

</details>

<details>

<summary>n=45k</summary>

```
Running test 10/10 for n = 45000...
Number of tokens in this prompt:  12072
Expected number in the prompt: 44988, mem output: 44988
Success!
Accuracy mem for n = 45000: 100.0%
```

</details>

<details>

<summary>n=50k</summary>

```
Running test 10/10 for n = 50000...
Number of tokens in this prompt:  13406
Expected number in the prompt: 31773, mem output: 1234
Fail.
Accuracy mem for n = 50000: 80.0%
```

</details>

<details>

<summary>n=60k</summary>

```
Running test 10/10 for n = 60000...
Number of tokens in this prompt:  16072
Expected number in the prompt: 31528, mem output: 31528
Success!
Accuracy mem for n = 60000: 100.0%
```

</details>

<details>

<summary>n=75k</summary>

```
Running test 10/10 for n = 75000...
Number of tokens in this prompt:  20071
Expected number in the prompt: 22901, mem output: 22901
Success!
Accuracy mem for n = 75000: 100.0%
```

</details>

Thanks to AutoMeta for compute for testing.

**Review by winglian — APPROVED** [review]
🔥

**Comment by NanoCode012** [comment]
This feature should allow longer context.

The test above uses the author's delta weights to confirm feature parity.

File: src/axolotl/utils/models.py [hunk 1]
```diff
@@ -83,37 +85,47 @@ def load_model(
...
         from axolotl.monkeypatch.llama_attn_hijack_xformers import (
             hijack_llama_sdp_attention,
         )
 
         logging.info("patching with sdp attention")
         hijack_llama_sdp_attention()
+    elif cfg.is_llama_derived_model and cfg.landmark_attention:
+        from axolotl.monkeypatch.llama_landmark_attn import (  # pylint: disable=redefined-outer-name # noqa: F811
+            MEM_TOKEN,
+            LlamaForCausalLM,
```
Comment (winglian) [inline src/axolotl/utils/models.py:122]: reassigning this here inside the function causes this error:
```
ERROR:root:local variable 'LlamaForCausalLM' referenced before assignment
Traceback (most recent call last):
  File "/workspace/axolotl/src/axolotl/utils/models.py", line 214, in load_model
    model = LlamaForCausalLM.from_pretrained(
UnboundLocalError: local variable 'LlamaForCausalLM' referenced before assignment
^CTraceback (most recent call last):
  File "/workspace/axolotl/src/axolotl/utils/models.py", line 214, in load_model
    model = LlamaForCausalLM.from_pretrained(
UnboundLocalError: local variable 'LlamaForCausalLM' referenced before assignment
```

Comment (NanoCode012) [reply src/axolotl/utils/models.py:122]: Hm, weirdly I didn't get this error.

Second, the new PR should fix this import issue. But, it hasn't been tested.

Comment (NanoCode012) [reply src/axolotl/utils/models.py:122]: I need to go afk for now, so I'll test the PR when I get back since it seem like a big bug.

**Pull request merged**